### DISCUSSION TITLE
Stepper Motors and LEDs with Asyncio

### DISCUSSION THREAD

**Discussion by Choons** [micropython/micropython#13582] [General]
Hi folks, my first post here, hoping someone can point me in the right direction. I'm building a moving head light that has addressable RGB LED  matrix(es) and two stepper motors to handle rotation. I want to be able to control it from a laptop over wifi + TCP socket. I've got the neopixel LED stuff working fine, stepper motors and TMC2209 controllers are sorted, and have my python app on laptop connecting to the ESP32 board with micropython over wifi socket into an asyncio socket server. I'm pretty new to asyncio coding and am looking for the best way to set up the async functions on the ESP32 board side to accept real time control for the stepper motors from say the arrow keys or even a joystick on the laptop side. Anyone done this or know of a code example? I'd rather not reinvent the wheel here. Thanks for any help.

**Comment by jimmo** [comment]
@Choons Is the idea that the laptop sends commands over WiFi? In which case you need to receive commands over a socket.

This might sound like overkill but a really simple way to do this is just to run an asyncio web server on your ESP32 and send commands over HTTP. I'd recommend taking a look at `microdot`.

This will make it super easy to write the client on the laptop because you can use any http client library (or even curl/wget from the command line!)

**Reply by Choons** [reply]
Hi Jim, yeah I guess the best way to think about it is imagine the laptop acting the same way as the controller for an RC plane. As close to real time as I can get. I have the wifi & socket setup running fine-- what I need help figuring out is the best way for the ESP32 to dispatch the received control commands to the stepper motor control function as well as to the RGB LEDs control function in a non-blocking way. I need the ESP32 to be able to handle doing both the fixture motion as well as the light patterns as close to concurrently as I can get. Asyncio seems like the right choice, but I haven't come across a code example that can handle real time input on a listening socket like that. On the laptop side, I am using a python program that sends raw bytes over TCP which works fine, so I'm not sure how using HTTP would be better? Mainly, I'm trying to wrap my head around calling the async functions on the ESP32 from the listening socket in the correct way.

**Reply by jimmo** [reply]
Mostly because something like microdot makes it trivial to route an incoming network request to an async function.

**Reply by Choons** [reply]
OK checking out microdot now. Thanks.

**Comment by peterhinch** [comment]
Asyncio is definitely the best way to achieve concurrency on a microcontroller. There is [a tutorial here](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md). To read from a socket concurrently with other tasks use a `StreamReader` as discussed [here](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md#63-using-the-stream-mechanism).

**Reply by Choons** [reply]
Ah OK, thanks much to you and Jim both for the code suggestions. I think I've got a grasp how to write it now.
